sb3_contrib/trpo/trpo.py

```diff
@@ -0,0 +1,387 @@
+                if isinstance(self.action_space, spaces.Discrete):
+                    # Convert discrete action from float to long
+                    actions = rollout_data.actions.long().flatten()
+
+                # Re-sample the noise matrix because the log_std has changed
+                # TODO: investigate why there is no issue with the gradient
+                # if that line is commented (as in SAC)
+                if self.use_sde:
+                    self.policy.reset_noise(self.batch_size)
+
+                with th.no_grad():
+                    old_distribution = copy.copy(self.policy.get_distribution(rollout_data.observations))
+
+                _, log_prob, _ = self.policy.evaluate_actions(rollout_data.observations, actions)
+                distribution = self.policy.action_dist
+
+                advantages = rollout_data.advantages
+                advantages = (advantages - advantages.mean()) / (rollout_data.advantages.std() + 1e-8)
+
+                # ratio between old and new policy, should be one at the first iteration
+                ratio = th.exp(log_prob - rollout_data.old_log_prob)
+
+                # surrogate policy objective
+                policy_obj = (advantages * ratio).mean()
+
+                # KL divergence
+                kl_div = kl_divergence(distribution.distribution, old_distribution.distribution).mean()
+
+                # Surrogate & KL gradient
+                self.policy.optimizer.zero_grad()
+
+                # This is necessary because not all the parameters in the policy have gradients w.r.t. the KL divergence
+                policy_obj_gradient = []
+                # Contains the gradients of the KL divergence
+                grad_kl = []
+                # Contains the shape of the gradients of the KL divergence w.r.t each parameter
+                # This way the flattened gradient can be reshaped back into the original shapes and applied to
+                # the parameters
+                grad_shape = []
+                # Contains the parameters which have non-zeros KL divergence gradients
+                # The list is used during the line-search to apply the step to each parameters
+                params = []
+
+                for param in self.policy.parameters():
+                    # For each parameter we compute the gradient of the KL divergence w.r.t to that parameter
+                    kl_param_grad, *_ = th.autograd.grad(
+                        kl_div,
+                        param,
+                        create_graph=True,
+                        retain_graph=True,
+                        allow_unused=True,
+                        only_inputs=True,
+                    )
+                    # If the gradient is not zero (not None), we store the parameter in the params list
+                    # and add the gradient and its shape to grad_kl and grad_shape respectively
+                    if kl_param_grad is not None:
```

Comment:
can that happen? (that `kl_param_grad` is None)
if so, in which case?

Reply:
yes - the Critic part of the Policy object that is independent of the Actor will have no gradients w.r.t to the KL divergence.

Reply:
thinking of this - is there a way to know, in the `ActorCriticPolicy` object, the Actor parameters at instantiation? That would speed up the code by avoiding calling `grad(kl_div, ...)` on the critic parameters.

Reply:
this would probably require a bit of work - since there are many different nets created at different steps (mlp extractor, SDE net, distribution net, value net). Is it worth it to create properties in ActorCriticPolicy returning the parameters for the actor, the critic and the feature net separately?

Reply:
so, I'm aware of the entanglement of the current implementation.
I recently created an alternative way of defining actor/critic for a course: https://github.com/osigaud/stable-baselines3/blob/master/stable_baselines3/reinforce/policies.py

I also plan to remove SDE net in the near future.

One way is to filter params by names, as done in https://stable-baselines3.readthedocs.io/en/master/guide/examples.html#accessing-and-modifying-model-parameters

>Is it worth it to create properties in ActorCriticPolicy returning the parameters for the actor, the critic and the feature net separately?

I think we need to create a discussion in SB3, where we will maybe remove this shared architecture in the future, or maybe allow to have a simpler actor/critic creation.
Anyway, I need to allow to have separate feature extractor (currently it is always shared between actor and critic).